test/plugin/test_metrics.rb

```diff
@@ -0,0 +1,294 @@
+
+  def initialize
+    super
+    @data = Hash.new(0)
+  end
+  def configure(conf)
+    super
+    class << self
+      alias_method :set, :set_counter
+    end
+  end
+  def get(key)
+    @data[key.to_s]
+  end
+  def inc(key)
+    @data[key.to_s] +=1
+  end
+  def add(key, value)
+    @data[key.to_s] += value
+  end
+  def set_counter(key, value)
+    @data[key.to_s] = value
+  end
+  def close
+    @data = {}
+    super
+  end
+end
+
+class BasicGaugeMetrics < Fluent::Plugin::Metrics
+  Fluent::Plugin.register_metrics('example', self)
+
+  attr_reader :data
+
+  def initialize
+    super
+    @data = Hash.new(0)
+  end
+  def get(key)
+    @data[key.to_s]
+  end
+  def inc(key)
+    @data[key.to_s] +=1
+  end
+  def dec(key)
+    @data[key.to_s] -=1
+  end
+  def add(key, value)
+    @data[key.to_s] += value
+  end
+  def sub(key, value)
+    @data[key.to_s] -= value
+  end
+  def set(key, value)
+    @data[key.to_s] = value
+  end
+  def close
+    @data = {}
+    super
+  end
+end
+
+class AliasedGaugeMetrics < Fluent::Plugin::Metrics
+  Fluent::Plugin.register_metrics('example', self)
+
+  attr_reader :data
+
+  def initialize
+    super
+    @data = Hash.new(0)
+  end
+  def configure(conf)
+    super
+    class << self
+      alias_method :dec, :dec_gauge
+      alias_method :set, :set_gauge
+      alias_method :sub, :sub_gauge
+    end
+  end
+  def get(key)
+    @data[key.to_s]
+  end
+  def inc(key)
+    @data[key.to_s] +=1
+  end
+  def dec_gauge(key)
+    @data[key.to_s] -=1
+  end
+  def add(key, value)
+    @data[key.to_s] += value
+  end
+  def sub_gauge(key, value)
+    @data[key.to_s] -= value
+  end
+  def set_gauge(key, value)
+    @data[key.to_s] = value
+  end
+  def close
+    @data = {}
+    super
+  end
+end
+
+class StorageTest < Test::Unit::TestCase
+  sub_test_case 'BareMetrics' do
+    setup do
+      @m = BareMetrics.new
+      @m.configure(config_element())
+    end
+
+    test 'is configured with plugin information and system config' do
+      m = BareMetrics.new
+      m.configure(config_element('metrics', '', {}))
+
+      assert_false m.use_gauge_metric
+      assert_false m.has_methods_for_counter
+      assert_false m.has_methods_for_gauge
+    end
+
+    test 'all operations are not defined yet' do
+      assert_raise NotImplementedError do
+        @m.get('key')
+      end
+      assert_raise NotImplementedError do
+        @m.inc('key')
+      end
+      assert_raise NotImplementedError do
+        @m.dec('key')
+      end
+      assert_raise NotImplementedError do
+        @m.add('key', 10)
+      end
+      assert_raise NotImplementedError do
+        @m.sub('key', 11)
+      end
+      assert_raise NotImplementedError do
+        @m.set('key', 123)
+      end
+    end
+  end
+
+  sub_test_case 'BasicCounterMetric' do
+    setup do
+      @m = BasicCounterMetrics.new
+      @m.configure(config_element('metrics', '', {'@id' => '1'}))
+    end
+
+    test 'all operations work well' do
```

Comment:
all "counter" operations?

Reply:
Like rspec context, "all operation work well" testcase should be interpret as `BasicCounterMetric` should `all operations work well`.
With your suggestion, it seems that it is redundant for "counter".
What do you think?